integrations/redis.go

```diff
@@ -0,0 +1,154 @@
+package integrations
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-redis/redis/v7"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+const (
+	fail    = "fail"
+	success = "success"
+)
+
+type redisContextKey int
+
+const (
+	redisCommandName redisContextKey = iota
+	redisCommandStartTime
+)
+
+func setRedisCommandStartTime(ctx context.Context) context.Context {
+	return context.WithValue(ctx, redisCommandStartTime, time.Now().UTC())
+}
+
+func getRedisCommandStartTime(ctx context.Context) (start time.Time) {
+	if t := ctx.Value(redisCommandStartTime); t != nil {
+		if startTime, ok := t.(time.Time); ok {
+			start = startTime
+			return
+		}
+	}
+	return
+}
+
+func setRedisCommandName(ctx context.Context, name string) context.Context {
+	return context.WithValue(ctx, redisCommandName, name)
+}
+
+func getRedisCommandName(ctx context.Context) (name string) {
+	if n := ctx.Value(redisCommandName); n != nil {
+		if cmdName, ok := n.(string); ok {
+			name = cmdName
+			return
+		}
+	}
+	return
+}
+
+// RedisBaseplateHook is a redis.Hook for wrapping Redis calls in client Spans
+// and metrics.
+type RedisBaseplateHook struct {
+	clientName string
+	metrics    metricsbp.Statsd
+}
+
+var _ redis.Hook = RedisBaseplateHook{}
+
+// NewRedisBaseplateHook creates a new RedisBaseplateHook to integrate a Redis
+// client from `go-redis` with Baseplate monitoring and tracing.
+func NewRedisBaseplateHook(ctx context.Context, clientName string, metricsConfig metricsbp.StatsdConfig) RedisBaseplateHook {
+	metrics := metricsbp.NewStatsd(ctx, metricsConfig)
```

Comment:
Every time you call `NewStatsd` a new background reporting goroutine will be started, so everything inside a process (micro service) should use a single `metricsbp.Statsd`. Here in `NewRedisBaseplateHook` we should just pass that in, instead of creating a new one.

For that reason we might not really need a constructor. We could just export the 2 fields so callers can just do:

```go
hook := integrations.RedisBaseplateHook{
  ClientName: "foo",
  Metrics: statsd,
}
```